### PR TITLE
Refactor VisitorInterface to allow for more fine-grained control

### LINKED ISSUE #42: Fix testMaintainsTypeInfoDuringEdit
One test is still failing from `VisitorTest`. The issue is most likely in the logic how we visit nodes added by visitors. I suspect that we are visited them in the wrong order.

### PR DESCRIPTION
**Pull request by Jalle19**
This started out as an attempt to fix #42 . Currently some tests are failing because "break" is not implemented.

### CODE REVIEW

**Comment by Jalle19** [comment]
Restored breaking functionality

File: src/Language/Visitor/AbstractReturnValue.php [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+
+use Digia\GraphQL\Language\Node\NodeInterface;
+
+abstract class AbstractReturnValue
+{
+
+    public const ACTION_NO_ACTION = 'NO_ACTION';
+    public const ACTION_BREAK     = 'BREAK';
+    public const ACTION_DELETE    = 'DELETE';
+    public const ACTION_REPLACE   = 'REPLACE';
```
Comment (crisu83) [inline src/Language/Visitor/AbstractReturnValue.php]: I would consider using an enumerable class for this, e.g. `VisitorActionEnum`.

File: src/Language/Visitor/EnterReturnValue.php [hunk 2]
```diff
@@ -0,0 +1,9 @@
+<?php
+
+namespace Digia\GraphQL\Language\Visitor;
+
+final class EnterReturnValue extends AbstractReturnValue
+{
+
+    public const ACTION_SKIP = 'SKIP';
```
Comment (crisu83) [inline src/Language/Visitor/EnterReturnValue.php]: What is this used for?

File: src/Language/Visitor/EnterReturnValue.php [hunk 3]
```diff
@@ -0,0 +1,9 @@
+<?php
+
+namespace Digia\GraphQL\Language\Visitor;
+
+final class EnterReturnValue extends AbstractReturnValue
```
Comment (crisu83) [inline src/Language/Visitor/EnterReturnValue.php]: Couldn't we just get by with one class, e.g. `VisitorResult`? If not, then I'd call these two classes `VisitorEnterResult` and `VisitorLeaveResult`.

**Comment by crisu83** [comment]
@Jalle19 This is not WIP anymore, right? If so, please remove the prefix from the title.

File: src/Language/Visitor/EnterReturnValue.php (hunk 2 above)
Comment (Jalle19) [reply src/Language/Visitor/EnterReturnValue.php]: They're not used anywhere, I just copied the values from the reference implementation

File: src/Language/Visitor/EnterReturnValue.php (hunk 3 above)
Comment (Jalle19) [reply src/Language/Visitor/EnterReturnValue.php]: I suppose we could, it's just that enter has one more possible value than leave

Comment (crisu83) [reply src/Language/Visitor/EnterReturnValue.php]: Ok fair enough, let's keep it like this then. 👍

**Comment by crisu83** [comment]
@Jalle19 Feel free to merge this. Thanks a lot for taking the time to refactor the visitor logic. 👍

**Comment by Jalle19** [comment]
I'm still thinking we could combine the two return value classes, and skip the unused constants

**Comment by crisu83** [comment]
What’s up w this PR @Jalle19?

**Comment by crisu83** [comment]
@Jalle19 I say we just merge this and do the rest of the refactoring when you have time.

**Comment by Jalle19** [comment]
I was actually gonna look at this today, so don't merge quite yet

**Comment by Jalle19** [comment]
@crisu83 updated

**Pull request merged**